File: backend/src/collectors/sslCollector.ts
import * as tls from "tls";

export interface SslResult {
  issuer: string | null;
  not_before: Date | null;
  not_after: Date | null;
  days_remaining: number | null;
  protocol: string | null;
  cipher: string | null;
  error: string | null;
}

const TIMEOUT_MS = 10_000;

export function checkSslCert(url: string): Promise<SslResult> {
  return new Promise((resolve) => {
    const parsed = new URL(url);
    const host = parsed.hostname;
    const port = parsed.port ? Number(parsed.port) : 443;

    const result: SslResult = {
      issuer: null,
      not_before: null,
      not_after: null,
      days_remaining: null,
      protocol: null,
      cipher: null,
      error: null,
    };

    const socket = tls.connect({ host, port, servername: host, timeout: TIMEOUT_MS }, () => {
      const cert = socket.getPeerCertificate();
      result.protocol = socket.getProtocol();
      result.cipher = socket.getCipher()?.name ?? null;
      const org = cert.issuer?.O ?? cert.issuer?.CN ?? null;
      result.issuer = Array.isArray(org) ? org[0] ?? null : org;

      if (cert.valid_from) result.not_before = new Date(cert.valid_from);
      if (cert.valid_to) {
        const notAfter = new Date(cert.valid_to);
        result.not_after = notAfter;
        result.days_remaining = Math.floor((notAfter.getTime() - Date.now()) / 86_400_000);
      }

      socket.end();
      resolve(result);
    });

    socket.once("timeout", () => {
      result.error = "connection timed out";
      socket.destroy();
      resolve(result);
    });
    socket.once("error", (err) => {
      result.error = err.message;
      resolve(result);
    });
  });
}

File: backend/app/collectors/ssl_collector.py
import socket
import ssl
from datetime import datetime
from urllib.parse import urlparse


def check_ssl_cert(url: str, timeout: float = 10.0) -> dict:
    parsed = urlparse(url)
    host = parsed.hostname
    port = parsed.port or 443

    result = {
        "issuer": None,
        "not_before": None,
        "not_after": None,
        "days_remaining": None,
        "protocol": None,
        "cipher": None,
        "error": None,
    }

    try:
        ctx = ssl.create_default_context()
        with socket.create_connection((host, port), timeout=timeout) as sock:
            with ctx.wrap_socket(sock, server_hostname=host) as tls_sock:
                cert = tls_sock.getpeercert()
                result["protocol"] = tls_sock.version()
                cipher = tls_sock.cipher()
                result["cipher"] = cipher[0] if cipher else None

        issuer_parts = dict(x[0] for x in cert.get("issuer", []))
        result["issuer"] = issuer_parts.get("organizationName") or issuer_parts.get("commonName")

        not_before = datetime.strptime(cert["notBefore"], "%b %d %H:%M:%S %Y %Z")
        not_after = datetime.strptime(cert["notAfter"], "%b %d %H:%M:%S %Y %Z")
        result["not_before"] = not_before
        result["not_after"] = not_after
        result["days_remaining"] = (not_after - datetime.utcnow()).days

    except Exception as exc:
        result["error"] = str(exc)[:512]

    return result
